src/server/project.ts

```diff
@@ -788,31 +764,39 @@ namespace ts.server {
                             trimmed = trimmed.substr(0, i);
                         }
                         (unresolvedImports || (unresolvedImports = [])).push(trimmed);
-                        result.push(trimmed);
                     }
                 });
             }
+
             this.cachedUnresolvedImportsPerFile.set(file.path, unresolvedImports || emptyArray);
+            return unresolvedImports || emptyArray;
 
             function isAmbientlyDeclaredModule(name: string) {
                 return ambientModules.some(m => m === name);
             }
         }
 
+        /* @internal */
+        setHasMoreOrLessFiles() {

```

Comment:
consider renaming the function to `onFileAddedOrRemoved` and the property to `hasAddedorRemovedFiles`, since the count can be the same, but we still need to invalidate the cache.